File: pilhas-filas/python/tickets.py
from collections import deque


def time_required_to_buy(tickets, k):
    queue = deque(range(len(tickets)))
    result = 0

    while True:
        i = queue.popleft()
        tickets[i] -= 1
        result += 1
        if tickets[i] > 0:
            queue.append(i)
        elif i == k:
            return result


print(time_required_to_buy([2, 3, 2], 2))
print(time_required_to_buy([5, 1, 1, 1], 0))

File: pilhas-filas/javascript/tickets.js
function timeRequiredToBuy(tickets, k) {
    let queue = [];
    let n = tickets.length;

    for (let i = 0; i < n; i++) {
        queue.push(i);
    }

    let time = 0;
    while (true) {
        let i = queue.shift();
        tickets[i]--;
        time++;
        if (tickets[i] > 0) {
            queue.push(i);
        } else if (i === k) {
            return time;
        }
    }
}

console.log(timeRequiredToBuy([2, 3, 2], 2));
console.log(timeRequiredToBuy([5, 1, 1, 1], 0));
